frontend/src/components/WeatherView.vue

```diff
@@ -0,0 +1,190 @@
+<script setup lang="ts">
+import { ref, onMounted, onUnmounted } from 'vue';
+import axios from 'axios';
+import wsClient from "@/utils/websocket";
+import type { Weather, User } from "@/types/ApiTypes";
+
+const API_DOMAIN = import.meta.env.VITE_API_DOMAIN;
+const users = ref<User[]>([]);
+const retryCounts = ref<{ [key: number]: number }>({});
+const apiError = ref(false);
+const refreshInterval = ref<number | null>(null);
+const emit = defineEmits<{
+  (e: "show-details", user: any): void;
+}>();
+
+let retryCount = 0;
+const maxRetries = 3;
+
+const fetchUsersData = async () => {
+  apiError.value = false;
+  try {
+    const response = await axios.get(`${API_DOMAIN}/users`);
+    users.value = response.data;
+
+    retryCount = 0;
+    await updateMissingWeather();
+  } catch (error) {
+    apiError.value = true;
+  }
+};
+
+const updateUserData = async (userIds: number[]) => {
+  const response = await axios.post(`${API_DOMAIN}/weather/update`, {
+    users: userIds,
+  });
+
+  const updatedUsers = response.data;
+  updatedUsers.forEach((updatedUser: User) => {
+    const index = users.value.findIndex((u: User) => u.id === updatedUser.id);
+    if (index !== -1 && updatedUser.weather) {
+      users.value[index].weather = updatedUser.weather;
+    }
+  });
+};
+
+const updateMissingWeather = async () => {
+  // Get IDs of users without weather data
+  const usersWithoutWeather = users.value
+    .filter((user: User) => !user.weather)
+    .map((user: User) => user.id);
+
+  if (usersWithoutWeather.length === 0) {
+    apiError.value = false;
+    return;
+  }
+
+  if (retryCount >= maxRetries) {
+    apiError.value = true;
+    return;
+  }
+
+  try {
+    await updateUserData(usersWithoutWeather);
```

Comment:
I like that you're doing retries here, but I find them a bit hard to follow with the way you've organized the logic. It might be helpful to pull this out into a composable. I think it's a bit tricky here since there are potentially 2 different endpoints that you are hitting but only one retryCount  and apiError. So it's hard to figure out which request could be failing.